### ISSUE TITLE
Feature Request: Make it possible to set a password for guests

### ISSUE THREAD

**Issue by Miterion** [hackmdio/hackmd#494]
We currently use etherpad with a custom pad manager so we can set a password to view/edit a pad. This way the links can be shared and not be edited by anyone. Would it be possible to add this to hackmd?

**Comment by SISheogorath** [comment]
You can use the usual auth integrations + permission.

Set it to limited and only users that are able to authenticate to a pad, can edit or see it.

**Comment by SISheogorath** [comment]
Interesting idea:

You can set **one** password for a pad in the permission section and everyone who would find a 403 right now, like Guest when a note is marked as  "Limited", "Protected" or "Private" or users in case of a "Private" one, will be prompted for the password. And additional checkbox, besides the password field should toggle write access to for these users.

It's a very basic feature, but should be enough.

**Comment by Miterion** [comment]
This would be very cool, would love to see this implemented or implement it myself

**Comment by SISheogorath** [comment]
We would very welcome a PR ❤️. Can't say when I'll get into this.

**Comment by SISheogorath** [comment]
From a UI perspective I did some experiments:

# Note implementation

## With label at the bottom:

![image](https://user-images.githubusercontent.com/8719867/31579090-a5a9213a-b12e-11e7-93b2-049e0496da58.png)

---

## Without at the bottom:

![image](https://user-images.githubusercontent.com/8719867/31579094-bf8fc252-b12e-11e7-9ace-7fa55e291f3b.png)

---

## With label but over the deletion in between:

![image](https://user-images.githubusercontent.com/8719867/31579105-06409e88-b12f-11e7-8f25-79537c2d3a93.png)

---

## Without label but over the deletion in between:

![image](https://user-images.githubusercontent.com/8719867/31579101-e741395c-b12e-11e7-8502-3152e5aca279.png)

---

# 403-page implementation

![screen shot 2017-10-14 at 22 43 47-fullpage](https://user-images.githubusercontent.com/8719867/31579364-fc1ec14a-b134-11e7-9e4f-9724675f5ade.png)


---

Right now, I didn't do any server-side implementation. So there is no branch for testing right now. I wonder what is preferred.

Any additional/alternative ideas?

**Comment by Miterion** [comment]
I would add a submit button to the 403 page, since many people would expect this.

**Comment by ccoenen** [comment]
regarding the different variants: I believe "note password" alone is not enough, it needs the extra label. Especially when something is already entered and the hint inside the input is no longer visible.

**Comment by SISheogorath** [comment]
Is this better:

![image](https://user-images.githubusercontent.com/8719867/31579745-051340c0-b13d-11e7-9f6e-cd450d49d721.png)

---

![image](https://user-images.githubusercontent.com/8719867/31579755-465b8b3c-b13d-11e7-9f59-2b1ac770d742.png)

---

## 403 with button

![screen shot 2017-10-15 at 00 32 20-fullpage](https://user-images.githubusercontent.com/8719867/31579868-53bab05c-b140-11e7-8fd4-759448396e93.png)

**Comment by SISheogorath** [comment]
Hints for the backend implementation: 

https://github.com/hackmdio/hackmd/blob/ca959012047bef3c0b2dd6f2974a8373a6615125/lib/response.js#L154-L158

https://github.com/hackmdio/hackmd/blob/ea9b1557f9834b2b140bd9b07488b5fca3953791/lib/realtime.js#L804-L811

https://github.com/hackmdio/hackmd/blob/ca959012047bef3c0b2dd6f2974a8373a6615125/lib/response.js#L20-L22

**Comment by Miterion** [comment]
I have a few questions regarding the backend implementation:
* When the password gets send to the server (from the 403 page), it has to be checked somewhere. I would implement it as a noteAction (see below)
https://github.com/hackmdio/hackmd/blob/ca959012047bef3c0b2dd6f2974a8373a6615125/lib/response.js#L366
Would that be okay or is there a better place
* My first idea for saving the password during a session is to save it with a cookie. The key would be the node id, the value the password.

Would love some feedback on these questions

**Comment by SISheogorath** [comment]
It's actually a good question how to store or save the password during this time. I didn't really think about that right now.

In general, we should do this by passport as every other auth before.

https://github.com/antgraf/passport-localapikey <-- seems to be good and maintained

Our note password would be the API key in this case. Means, we need to implement the authentication strategy but not for the usual user accounts. This answers your question where to send the password. The usual `/auth/<something>` URL. I guess `/auth/api` would be fine.

I'm not sure if this works, but that's my idea for now.

**Comment by Miterion** [comment]
I had an idea on how to make this without another authentication mechanism:

When a user sends the correct password for a note to the `auth` system the note id could be saved in the users session (when the password is correct). This way, the system only has to check whether the requested note id is in this session variable and grant/deny access to it.

**Comment by SISheogorath** [comment]
Sounds fine, too. Can you provide some example code? I'm not completely sure how the implementation looks like, feel free to surprise me ^^

**Comment by SISheogorath** [comment]
If you can make it to land a PR until end of next week I'll review and merge it to 1.0.0-CE. Otherwise this Feature goes for next release

**Comment by Miterion** [comment]
Please push it back to a later release, have a lot assignements at university right now

**Comment by SISheogorath** [comment]
@Miterion hey, how is it going? Still interested in doing this?

**Comment by Miterion** [comment]
@SISheogorath I do not think that I am able to work on this currently. If someone else wants to do it I would be really happy

**Comment by ccoenen** [comment]
I feel this is also related to #138, perhaps this is a way to have some notes encrypted?

**Comment by mgoldau** [comment]
Hi, are there any update on this feature, to protect certain pads by password? By the way, the current permissions, are mostly concerning edit permissions, only "private" states that only myself can view the pad. However the usecase for me is that I want to have a private pad, where only users with the password can view and edit, regardless if they authenticated/logged in or not. As I am using the service in a mass-installation (gwdg), I don't want to share some sensitive pads with the whole community.

**Comment by mz83ude** [comment]
> 
> 
> Hi, are there any update on this feature, to protect certain pads by password? By the way, the current permissions, are mostly concerning edit permissions, only "private" states that only myself can view the pad. However the usecase for me is that I want to have a private pad, where only users with the password can view and edit, regardless if they authenticated/logged in or not. As I am using the service in a mass-installation (gwdg), I don't want to share some sensitive pads with the whole community.


**Absolutely agree with the suggestion!**

This is a necessary feature, especially in large installations (k-range).

**Comment by janit42** [comment]
+1

**Comment by alexpovel** [comment]
Also expressing interest in this.

**Comment by VaalaCat** [comment]
This feature is really necessary for me

**Comment by asakura42** [comment]
+1

**Comment by jmsjsph** [comment]
+1

**Comment by zubair1024** [comment]
+1
